### PR TITLE
Pose SLAM initialization

### PR DESCRIPTION
**Pull request by akshay-krishnan**

### CODE REVIEW

File: gtsfm/pose_slam/pose_slam.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import dask
+import gtsam
+from dask.delayed import Delayed
+from gtsam import Pose3
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.pose_prior import PosePrior
+from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
+
+PRIOR_NOISE_SIGMAS = [0.001, 0.001, 0.001, 0.1, 0.1, 0.1]
```
Comment (ayushbaid) [inline gtsfm/pose_slam/pose_slam.py:16]: super strict on rotations?

File: gtsfm/pose_slam/pose_slam.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+POSE_PRIOR_NOISE = gtsam.noiseModel.Diagonal.Sigmas(PRIOR_NOISE_SIGMAS)
+
+logger = logger_utils.get_logger()
+
+
+class PoseSlam:
+    def run_pose_slam(
+        self,
+        num_images: int,
+        relative_pose_priors: Dict[Tuple[int, int], PosePrior] = {},
```
Comment (ayushbaid) [inline gtsfm/pose_slam/pose_slam.py:26]: Optional arguments are dangerous. Also, we would have empty list if loader doesnt have priors.

File: gtsfm/pose_slam/pose_slam.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                    gtsam.noiseModel.Gaussian.Covariance(i1Ti2_prior.covariance),
+                )
+            )
+        pose_init_graph.push_back(gtsam.PriorFactorPose3(0, Pose3(), POSE_PRIOR_NOISE))
+
+        initial_values = gtsam.InitializePose3.initialize(pose_init_graph)
+        initial_error = pose_init_graph.error(initial_values)
+        logger.info(f"[pose slam] Pose SLAM initialization complete with error: {initial_error}")
+        optimizer = gtsam.LevenbergMarquardtOptimizer(pose_init_graph, initial_values)
+        result = optimizer.optimizeSafely()
```
Comment (ayushbaid) [inline gtsfm/pose_slam/pose_slam.py:58]: What does this do?

File: gtsfm/pose_slam/pose_slam.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        poses = [result.atPose3(i) if result.exists(i) else None for i in range(num_images)]
+        final_error = pose_init_graph.error(result)
+        logger.info(f"[pose slam] Pose SLAM optimization complete with error: {final_error}")
+
+        return poses, GtsfmMetricsGroup(
+            "pose_slam_metrics", [GtsfmMetric("intial_error", initial_error), GtsfmMetric("final_error", final_error)]
```
Comment (ayushbaid) [inline gtsfm/pose_slam/pose_slam.py:65]: Can you compare the global poses with ground truth? We have a function for this: geometry_comparison.compare_global_poses. And maybe other functions over there if needed.

File: gtsfm/multi_view_optimizer.py [hunk 5]
```diff
@@ -24,11 +25,14 @@ def __init__(
         trans_avg_module: TranslationAveragingBase,
         data_association_module: DataAssociation,
         bundle_adjustment_module: BundleAdjustmentOptimizer,
+        use_pose_slam_intitialization: bool = False,
     ) -> None:
+        self.pose_slam_module = PoseSlam()
```
Comment (ayushbaid) [inline gtsfm/multi_view_optimizer.py:30]: nit: It should go into hydra config later?

File: gtsfm/configs/deep_front_end_hilti.yaml [hunk 6]
```diff
@@ -69,6 +69,8 @@ SceneOptimizer:
       output_reproj_error_thresh: 3 # for post-optimization filtering
       robust_measurement_noise: True
       shared_calib: False
+    
+    use_pose_slam_intitialization: True
```
Comment (ayushbaid) [inline gtsfm/configs/deep_front_end_hilti.yaml:73]: Can you please move it up with the target for multi-view plate? In both the configs?

**Review by ayushbaid — APPROVED** [review]

File: gtsfm/pose_slam/pose_slam.py (hunk 1 above)
Comment (dellaert) [reply gtsfm/pose_slam/pose_slam.py:16]: It is arbitrary. This prior will be satisfied exactly

File: gtsfm/pose_slam/pose_slam.py (hunk 3 above)
Comment (dellaert) [reply gtsfm/pose_slam/pose_slam.py:58]: I forget :-) doxygen…

File: gtsfm/pose_slam/pose_slam.py (hunk 4 above)
Comment (dellaert) [reply gtsfm/pose_slam/pose_slam.py:65]: Could we leave this for later?

File: gtsfm/multi_view_optimizer.py (hunk 5 above)
Comment (dellaert) [reply gtsfm/multi_view_optimizer.py:30]: There is no alternative. When we have two versions, yes

**Comment by dellaert** [comment]
@akshay-krishnan you’re addressing unresolved comments and merging right?

**Comment by akshay-krishnan** [comment]
> @akshay-krishnan you’re addressing unresolved comments and merging right?

yes, doing that now.

**Comment by ayushbaid** [comment]
I'll push the runner file with the modifications to this PR too.

**Pull request merged**